### PR TITLE
chore!: Update hugr to `0.8.0`

### LINKED ISSUE #410: Add custom extension with Partial Application op
See https://github.com/CQCL/guppylang/blob/621b44947513bbc6b29887b7e0e4ea49e045a3ec/guppylang/compiler/func_compiler.py#L116-L130

### LINKED ISSUE #411: Replace `guppylang.unsupported` ops with new hugr definitions
https://github.com/CQCL/hugr/pull/1463 adds a number of ops we were missing here, we should use those instead of unsuported opaque ops.

### LINKED ISSUE #412: Use extOps definitions from hugr-py 0.8
`hugr-py 0.8` will export all the std operations defined as as extension ops.
(Backed by a single centralised extension definition).
https://github.com/CQCL/hugr/pull/1459

We should be able to replace the hacky opaque op definitions in `.prelude._internal.util`/`.prelude.builtins` with direct references to the operations.

### LINKED ISSUE #442: Update to hugr 0.8
This is a prerequisite for most issues in https://github.com/CQCL/guppylang/issues/391

### PR DESCRIPTION
**Pull request by aborgna-q**
Updates guppy to use extension operation definitions instead of opaque ops defined on the fly.

Some relevant changes included:

* Defined a new `guppylang` extension with two ops: `partial` and `unsupported`
* Most list and array operations currently emit `unsupported`, as they require custom hugr building code (the core ops in `std.collections` don't match the python operations 1 to 1, it'll need some custom compilation code).
* Changed `module.compile` to produce a `Package` containing the hugr along with the required extensions to interpret it. (Currently hardcoded, we should infer which extensions are needed at some point).
* We hardcode the tket2 extension jsons. The plan is for `tket2` to publish a new library exporting those, but for now we need the jsons for validation.
* Use the prelude operations exported by hugr for most int/float/logic/conversion builtin definitions.

The `tket2` dependency is currently pinned to a commit hash, I'll update it once `0.3.0` is released.

- Closes #442
- Closes #411
- Closes #412
- Closes #410 

BREAKING CHANGE: Bumped the `hugr` dependency to `0.8.0`

### CODE REVIEW

File: guppylang/prelude/builtins.py [hunk 1]
```diff
@@ -194,29 +193,29 @@ def __or__(self: nat, other: nat) -> nat: ...
...
     def __pos__(self: nat) -> nat: ...
 
-    @guppy.hugr_op(builtins, custom_op("ipow", args=[int_arg()]))  # TODO
+    @guppy.hugr_op(builtins, int_op("ipow"))
     def __pow__(self: nat, other: nat) -> nat: ...
 
     @guppy.hugr_op(builtins, int_op("iadd"), ReversingChecker())
     def __radd__(self: nat, other: nat) -> nat: ...
 
-    @guppy.hugr_op(builtins, int_op("rand"), ReversingChecker())
```
Comment (aborgna-q) [inline guppylang/prelude/builtins.py:203]: An upside of this change is that we can now detect these mislabels.

**Review by mark-koch — COMMENTED** [review]
Nice! 🎉

Just a few nits

**Review by mark-koch — APPROVED** [review]

File: guppylang/prelude/_internal/compiler/arithmetic.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    op_def = hugr.std.int.CONVERSIONS_EXTENSION.get_op("itobool")
+    return ops.ExtOp(
+        op_def,
+        ht.FunctionType([int_t(1)], [ht.Bool]),
+    )
+
+
+# ------------------------------------------------------
+# --------- Custom compilers for non-native ops --------
+# ------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/prelude/_internal/compiler/arithmetic.py:91]: Yes, let's avoid bloating this PR even more. I'll make the issue

File: guppylang/prelude/_internal/compiler/prelude.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+# ------------------------------------------------------
+# --------- Custom compilers for non-native ops --------
+# ------------------------------------------------------
+
+
+def build_panic(
+    # TODO: Change to `DfBase[ops.DfParentOp]` once `DfBase` is covariant
+    builder: DfBase[ops.Case],
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/prelude/_internal/compiler/prelude.py:63]: No, and it won't be covariant.
I'll drop the TODO

File: guppylang/prelude/_internal/compiler/quantum.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    QUANTUM_EXTENSION,
+    RESULT_EXTENSION,
+]
+
+
+ANGLE_T = QUANTUM_EXTENSION.get_type("angle")
+
+# ------------------------------------------------------
+# --------- Custom compilers for non-native ops --------
+# ------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (aborgna-q) [reply guppylang/prelude/_internal/compiler/quantum.py:35]: #464

**Pull request merged**